### DISCUSSION TITLE
I’m facing a database query error while fetching event data

### DISCUSSION THREAD

**Discussion by ajselvan** [aptos-labs/aptos-developer-discussions#575] [Questions]
### Describe your question in detail.

```
query MyQuery {
  events(
    where: {data: {_contains: {user_address: "0x829faddf57a5a6c9f1563e5e3c643f9942039486a9b030517187c7baebd4fc87"}}}
    order_by: {transaction_version: desc}
    limit: 2
  ) {
    transaction_version
    data
    type
  }
}
```
This is my query code to fetch the events. The default limit is 100, right?

When I provide a limit of 100, I get a database query error, even though I only have 2 event entries. If I set the limit to 2, it returns the response correctly.

However, if I set a limit greater than the available event data, it returns an error. The issue is that we don't know the exact length of the event data. Could you please help me resolve this? Thank you!

### What error, if any, are you getting?

![Screenshot (506)](https://github.com/user-attachments/assets/e3d23c92-93e9-4045-b698-5ee98f54ef7d)

Here is my query code & error screenshot

### What have you tried or looked at?  Or how can we reproduce the error?

Please try running the query code I sent at the top. The default limit is 100, so you are getting a query error. However, if you set the limit to exactly 2, it will return the event data. This is because I only have 2 events for that address.

This is my example code:

```
import { AptosConfig, Network, Aptos } from "@aptos-labs/ts-sdk";


async function main(): Promise<void> {

    const config = new AptosConfig({ network: Network.TESTNET });
    const aptos = new Aptos(config);

    const getAllEvents: any = await aptos.queryIndexer({
        query: {
            query: `
        query MyQuery {
        events(
            where: {data: {_contains: {user_address: "0x829faddf57a5a6c9f1563e5e3c643f9942039486a9b030517187c7baebd4fc87"}}}
            order_by: {transaction_version: desc}
            limit: 2
        ) {
         transaction_version
         data
         type
        }
    }
`
        }
    })
    console.log("getAllEvents", getAllEvents.events);

}


main().catch(error => {
    console.error('An error occurred:', error);
});
```

### Which operating system are you using?

Windows

### Which SDK or tool are you using? (if any)

TypeScript SDK

**Comment by ying-w** [comment]
searching in `data` is expensive. Ideally you emit event from user_address (so you can filter on event account_address)

You can also get additional quota (not sure if it will solve this timeout problem) by getting API key through aptos labs developer portal https://developers.aptoslabs.com/docs/api-access/api-keys

**Comment by 0x-j** [comment]
you can also run your own indexer easily so you only index events you are interested in. 
here's a template to run your own indexer 
https://github.com/0xaptosj/aptos-full-stack-template
